lib/Raven/ErrorHandler.php

```diff
@@ -11,203 +9,331 @@
-        if (PHP_VERSION_ID >= 70000 && E_ERROR === $type) {
-            return false;
+        if (!empty($error) && $error['type'] & (E_ERROR | E_PARSE | E_CORE_ERROR | E_CORE_WARNING | E_COMPILE_ERROR | E_COMPILE_WARNING)) {
+            $exception = new \ErrorException(self::$errorLevelsDescriptions[$error['type']] . ': ' . $error['message'], 0, $error['type'], $error['file'], $error['line']);
         }
 
-        return $type & $this->fatal_error_types;
+        try {
+            if (null !== $exception) {
+                $this->handleException($exception);
+            }
+        } catch (\ErrorException $exception) {
+            // Ignore this re-throw
+        }
     }
 
     /**
-     * Register a handler which will intercept unhandled exceptions and report them to the
-     * associated Sentry client.
+     * Handles the given exception by capturing it through the Raven client and
+     * then forwarding it to another handler.
      *
-     * @param bool $call_existing Call any existing exception handlers after processing this instance
+     * @param \Exception|\Throwable $exception The exception to handle
      *
-     * @return \Raven\ErrorHandler
+     * @throws \Exception|\Throwable
+     *
+     * @internal
      */
-    public function registerExceptionHandler($call_existing = true)
+    public function handleException($exception)
     {
-        $this->old_exception_handler = set_exception_handler([$this, 'handleException']);
-        $this->call_existing_exception_handler = $call_existing;
+        $this->client->captureException($exception);
+
+        // If there is no exception handler that can run after the current one
+        // give back the exception to the native PHP handler
+        if (null === $this->previousExceptionHandler) {
+            throw $exception;
+        }
+
+        $previousExceptionHandlerException = $exception;
+
+        // Unset the previous exception handler to prevent infinite loop in case
+        // we need to handle an exception thrown from it
+        $previousExceptionHandler = $this->previousExceptionHandler;
+        $this->previousExceptionHandler = null;
+
+        try {
+            call_user_func($previousExceptionHandler, $exception);
+        } catch (\Exception $previousExceptionHandlerException) {
+            // Do nothing, we just need to set the $previousExceptionHandlerException
+            // variable to the exception we just catched to compare it later
+            // with the original object instance
+        }
+
+        // If the exception object instance is the same as the one catched from
+        // the previous exception handler, if any, give it back to the native
+        // PHP handler to prevent infinite circular loop
+        if ($exception === $previousExceptionHandlerException) {
+            throw $exception;
+        }
 
-        return $this;
+        $this->handleException($previousExceptionHandlerException);
```

Comment:
Is this loop-prone?

Reply:
Yes it should be as the next time the function will be called `$this->previousExceptionHandler` will be `null` and after sending the exception to Sentry the exception itself will be rethrown